### DISCUSSION TITLE
compiling error on RPi4 esphome 2023.3

### DISCUSSION THREAD

**Discussion by the-butterfry** [syssi/esphome-jbd-bms#35] [Q&A]
Just trying to deploy a new BMS on a esphome + HA coupled Rpi 4

I am able to compile and deploy fine on x86, but not this rpi. Looks like you're messing around with the version, as I'm sure you're aware @syssi latest doesn't work anymore. The following is from Recommended:

```
INFO Reading configuration /config/esphome/jbd-bms-farm.yaml...
INFO Updating https://github.com/syssi/esphome-jbd-bms.git@main
INFO Generating C++ source...
INFO Compiling app...
Processing jbd-bms-farm (board: esp32doit-devkit-v1; framework: espidf; platform: platformio/espressif32 @ 5.3.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40404.0 (4.4.4) 
 - tool-cmake @ 3.16.9 
 - tool-ninja @ 1.10.2 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Reading CMake configuration...
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc -- broken
-- Configuring incomplete, errors occurred!
See also "/data/jbd-bms-farm/.pioenvs/jbd-bms-farm/CMakeFiles/CMakeOutput.log".
See also "/data/jbd-bms-farm/.pioenvs/jbd-bms-farm/CMakeFiles/CMakeError.log".

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/tool-cmake/share/cmake-3.16/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc"
```

  is not able to compile a simple test program.

  It fails with the following output:

```
    Change Dir: /data/jbd-bms-farm/.pioenvs/jbd-bms-farm/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/data/cache/platformio/packages/tool-ninja/ninja cmTC_ccea4 && [1/2] Building C object CMakeFiles/cmTC_ccea4.dir/testCCompiler.c.obj
    [2/2] Linking C executable cmTC_ccea4
    FAILED: cmTC_ccea4 
    : && /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc -mlongcalls -Wno-frame-address   CMakeFiles/cmTC_ccea4.dir/testCCompiler.c.obj  -o cmTC_ccea4   && :
    /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: error loading plugin: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: cannot open shared object file: No such file or directory
    collect2: error: ld returned 1 exit status
    ninja: build stopped: subcommand failed.

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf@3.40404.0/tools/cmake/project.cmake:296 (__project)
  CMakeLists.txt:3 (project)
```

**Comment by syssi** [comment]
Please remove the platform `version: latest` from your YAML and try again.

**Comment by the-butterfry** [comment]
I did - that compile log is from no version specified or the recommended

**Reply by syssi** [reply]
Could you try to build this YAML (with touching the ESPHome name): https://raw.githubusercontent.com/syssi/esphome-jbd-bms/main/esp32-ble-scanner.yaml

**Reply by the-butterfry** [reply]
Same:

INFO Reading configuration /config/esphome/bms-test.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing bms-test (board: wemos_d1_mini32; framework: espidf; platform: platformio/espressif32 @ 5.3.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40404.0 (4.4.4) 
 - tool-cmake @ 3.16.9 
 - tool-ninja @ 1.10.2 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Reading CMake configuration...
-- Found Git: /usr/bin/git (found version "2.30.2") 
-- The C compiler identification is GNU 8.4.0
-- The CXX compiler identification is GNU 8.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc -- broken
-- Configuring incomplete, errors occurred!
See also "/data/bms-test/.pioenvs/bms-test/CMakeFiles/CMakeOutput.log".
See also "/data/bms-test/.pioenvs/bms-test/CMakeFiles/CMakeError.log".

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/tool-cmake/share/cmake-3.16/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/bms-test/.pioenvs/bms-test/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/data/cache/platformio/packages/tool-ninja/ninja cmTC_09817 && [1/2] Building C object CMakeFiles/cmTC_09817.dir/testCCompiler.c.obj
    [2/2] Linking C executable cmTC_09817
    FAILED: cmTC_09817 
    : && /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/xtensa-esp32-elf-gcc -mlongcalls -Wno-frame-address   CMakeFiles/cmTC_09817.dir/testCCompiler.c.obj  -o cmTC_09817   && :
    /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: error loading plugin: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: cannot open shared object file: No such file or directory
    collect2: error: ld returned 1 exit status
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf@3.40404.0/tools/cmake/project.cmake:296 (__project)
  CMakeLists.txt:3 (project)



========================== [FAILED] Took 9.39 seconds ==========================

**Reply by the-butterfry** [reply]
here was my config:

substitutions:
  name: bms-test
  device_description: "Search for BLE devices"

esphome:
  name: ${name}
  comment: ${device_description}
  project:
    name: "syssi.esphome-jbd-bms"
    version: 1.4.0

esp32:
  board: wemos_d1_mini32
  framework:
    type: esp-idf

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

ota:

logger:
  level: DEBUG
  logs:
    wifi: WARN

api:
  reboot_timeout: 0s

esp32_ble_tracker:
  on_ble_advertise:
    then:
      - lambda: |-
          if (x.get_name().rfind("xiaoxiang", 0) == 0) {
            ESP_LOGI("ble_adv", "New JBD-BMS found");
            ESP_LOGI("ble_adv", "  Name: %s", x.get_name().c_str());
            ESP_LOGI("ble_adv", "  MAC address: %s", x.address_str().c_str());
            ESP_LOGD("ble_adv", "  Advertised service UUIDs:");
            for (auto uuid : x.get_service_uuids()) {
              ESP_LOGD("ble_adv", "    - %s", uuid.to_string().c_str());
            }
          }

text_sensor:
  - platform: ble_scanner
    name: "BLE Devices Scanner"

**Reply by syssi** [reply]
That's a pity. It looks like the build toolchain on a Raspberry Pi 4 is incompatible or just broken. This shouldn't happen:

```
/data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: error loading plugin: /data/cache/platformio/packages/toolchain-xtensa-esp32@8.4.0+2021r2-patch5/bin/../libexec/gcc/xtensa-esp32-elf/8.4.0/liblto_plugin.so: cannot open shared object file: No such file or directory
```

Do you have access to another linux host?

**Reply by the-butterfry** [reply]
yeah - another ARM? I think so, let me see.

 Like i said originally, thinking i was going crazy - i tried it on an x86 debian / HA install and it was fine once I switched to recommended.

**Reply by the-butterfry** [reply]
I'll try a new HA OS install - The one that is failing is a few years old and has some quirks to it.

**Reply by syssi** [reply]
I could provide some guidance how to build the ESPHome node on a x86 debian system:

![Screencast](https://raw.githubusercontent.com/syssi/esphome-seplos-bms/main/install.gif)

**Reply by the-butterfry** [reply]
the problem isn't x86 - I can confirm that works fine. It's ARM RPI based that was failing.

I just did a new HAOS install on another RPI and it's passing now. Seems to be something with that particular install. 
Linux cosmic-farm 5.10.103-v7l+ #1529 SMP Tue Mar 8 12:24:00 GMT 2022 armv7l

**Reply by syssi** [reply]
Good to know ARM is working in general.
